Networked 'dora build --uv' + 'dora start --uv' fails: managed Python env not recorded for start (only 'dora run' works)

## Summary

For a Python node using a managed Python env (`--uv`), the networked flow `dora build --uv` + `dora start --uv` fails at start with "no managed env was recorded for it during build", even though `dora build` succeeded and created the env on disk. The single-process `dora run --uv` works, so only the separate-invocation (build then start) path is broken.

## Reproduction (manual, no test harness)

```
$ dora up
$ dora build examples/python-async/dataflow.yaml --uv      # exit 0
   send_data: stdout   Using Python 3.13.5 environment at: .dora/python-envs/send_data
   send_data: INFO     running build command: `pip install numpy pyarrow` ...
$ dora start examples/python-async/dataflow.yaml --detach --uv   # exit 1
   [ERROR] node `receive_data_with_sleep` is a Python node that needs a managed env
   under `--uv`, but no managed env was recorded for it during build. Re-run
   `dora build --uv <dataflow>` against the current build session, or omit `--uv`
   to run against the ambient Python.
   Location: binaries/daemon/src/lib.rs:2671:25
```

`dora build` clearly DID create the per-node managed envs under `.dora/python-envs/<node>/`, but `dora start` (a separate invocation) cannot find them recorded for the build session — so the build→start handoff of managed-env state is lost.

Contrast: `dora run examples/python-async/dataflow.yaml --uv --stop-after 10s` (single process, build+run together) works fine.

## Impact

Release-relevant for 1.0.0-rc1: networked `dora up` + `dora build --uv` + `dora start --uv` is the primary multi-process deployment path, and it's broken for every Python node with a managed env. It's masked in CI because the nightly example harness runs these examples via `dora run --uv` (local single-process), e.g. `scripts/qa/ci-nightly-jobs.sh` runs `dora run examples/python-async/dataflow.yaml --uv` — never the networked build→start path.

Reproduces for: python-async, python-drain, python-log, python-multiple-arrays, python-concurrent-rw, python-zero-copy-send (+ contract) — every managed-env example. (Found alongside #2003; #2003's `uv build --clear` bug previously masked this by failing at the build step.)

## Pointers

- `binaries/daemon/src/lib.rs:2671` — the "no managed env was recorded" error.
- Managed-env feature: #1820 / #1829.
- The build session / build_id handoff between `dora build` and a later `dora start` (the on-disk `out/<name>.dora-session.yaml` and how managed envs are keyed to it).

## Environment

uv 0.8.3, macOS arm64, workspace 1.0.0-rc1, with the #2003 fix applied (so build succeeds).

--- a/binaries/daemon/src/lib.rs
+++ b/binaries/daemon/src/lib.rs
@@ -2466,6 +2466,10 @@ impl Daemon {
 
         let mut stopped = Vec::new();
 
+        // A present `build_id` means the session has a build that wasn't
+        // invalidated (the CLI clears it when build inputs change). We use this
+        // below to decide whether re-deriving an on-disk managed env is safe.
+        let have_build_id = build_id.is_some();
         let build_info = build_id.and_then(|build_id| self.builds.get(&build_id));
         let node_with_git_source = nodes.values().find(|n| n.has_git_source());
         if let Some(git_node) = node_with_git_source
@@ -2664,24 +2668,29 @@ impl Daemon {
                 // process from the daemon that later serves `dora start`, so the
                 // daemon's in-memory build record can be empty (dora-rs/dora#2004).
                 // The env dir is deterministic, so re-derive it (the restart path
-                // does the same), and use it IF `dora build` actually created it on
-                // disk. Fail closed only when the env is genuinely missing (stale
-                // session, prior non-`--uv` build, or `start --uv` without a build)
-                // so `--uv` never silently falls back to the ambient Python.
+                // does the same) and reuse the on-disk env -- but ONLY when a
+                // build id is present, i.e. a build occurred and this daemon just
+                // lacks the in-memory record. When the build id was cleared (the
+                // CLI invalidates it on build-input changes, a prior non-`--uv`
+                // build, or `start --uv` with no build), do NOT reuse a possibly
+                // stale env; require a rebuild. Either way, never silently fall
+                // back to the ambient Python.
                 if uv
                     && configured_python_env_dir.is_none()
                     && let Some(expected) =
                         dora_core::build::managed_python_env_dir(&node, &node_working_dir)
                 {
-                    if dora_core::build::managed_python_interpreter(&expected).is_file() {
+                    if have_build_id
+                        && dora_core::build::managed_python_interpreter(&expected).is_file()
+                    {
                         configured_python_env_dir = Some(expected);
                     } else {
                         eyre::bail!(
                             "node `{node_id}` is a Python node that needs a managed env under `--uv`, \
-                             but none was found at `{}`. \
+                             but no current build provides one (the build cache is absent or was \
+                             invalidated by changed build inputs). \
                              Run `dora build --uv <dataflow>` before `dora start --uv`, \
-                             or omit `--uv` to run against the ambient Python.",
-                            expected.display()
+                             or omit `--uv` to run against the ambient Python."
                         );
                     }
                 }

--- a/binaries/daemon/src/lib.rs
+++ b/binaries/daemon/src/lib.rs
@@ -2658,21 +2658,30 @@ impl Daemon {
                 let node_write_events_to = write_events_to
                     .as_ref()
                     .map(|p| p.join(format!("inputs-{}.json", node.id)));
-                let configured_python_env_dir = python_env_dirs.get(&node_id).cloned();
-                // Fail closed under `--uv` when the build artifacts don't include a
-                // managed env for a node that needs one — happens with a stale session
-                // file, a prior non--uv build, or `dora start --uv` without rebuilding.
-                // Falling back silently would lose the isolation guarantees the user
-                // asked for by passing `--uv`.
-                if uv && configured_python_env_dir.is_none() {
-                    let expected =
-                        dora_core::build::managed_python_env_dir(&node, &node_working_dir);
-                    if expected.is_some() {
+                let mut configured_python_env_dir = python_env_dirs.get(&node_id).cloned();
+                // Under `--uv`, a node may have no recorded managed env even when
+                // `dora build` prepared one: a networked build runs in a separate
+                // process from the daemon that later serves `dora start`, so the
+                // daemon's in-memory build record can be empty (dora-rs/dora#2004).
+                // The env dir is deterministic, so re-derive it (the restart path
+                // does the same), and use it IF `dora build` actually created it on
+                // disk. Fail closed only when the env is genuinely missing (stale
+                // session, prior non-`--uv` build, or `start --uv` without a build)
+                // so `--uv` never silently falls back to the ambient Python.
+                if uv
+                    && configured_python_env_dir.is_none()
+                    && let Some(expected) =
+                        dora_core::build::managed_python_env_dir(&node, &node_working_dir)
+                {
+                    if dora_core::build::managed_python_interpreter(&expected).is_file() {
+                        configured_python_env_dir = Some(expected);
+                    } else {
                         eyre::bail!(
                             "node `{node_id}` is a Python node that needs a managed env under `--uv`, \
-                             but no managed env was recorded for it during build. \
-                             Re-run `dora build --uv <dataflow>` against the current build session, \
-                             or omit `--uv` to run against the ambient Python."
+                             but none was found at `{}`. \
+                             Run `dora build --uv <dataflow>` before `dora start --uv`, \
+                             or omit `--uv` to run against the ambient Python.",
+                            expected.display()
                         );
                     }
                 }

